Clean/Table/Basic.lean

```diff
@@ -125,249 +125,249 @@ end TraceOfLength
   To make sure that the vertical offset is bounded, it is represented as a `Fin W`.
 -/
 structure CellOffset (W: ℕ+) (S : Type → Type) [ProvableType S]  where
-  rowOffset: Fin W
+  row: Fin W
   column: Fin (size S)
-deriving Repr
+
+instance : Repr (CellOffset W S) where
+  reprPrec off _ := "⟨" ++ reprStr off.row ++ ", " ++ reprStr off.column ++ "⟩"
 
 namespace CellOffset
 
 /--
   Current row offset
 -/
-@[table_norm]
+@[table_assignment_norm]
 def curr {W : ℕ+} (j : Fin (size S)) :  CellOffset W S := ⟨0, j⟩
 
 /--
   Next row offset
 -/
-@[table_norm]
+@[table_assignment_norm]
 def next {W : ℕ+} (j : Fin (size S)) :  CellOffset W S := ⟨1, j⟩
 
 end CellOffset
 
-/--
-  Mapping from the index of a variable to a cell offset in the table.
--/
-@[reducible]
-def CellAssignment (W: ℕ+) (S : Type → Type) [ProvableType S] := ℕ → CellOffset W S
-
-/--
-  Atomic operations for constructing a table constraint, which is a constraint applied to a window
-  of rows in a table.
--/
-inductive TableConstraintOperation (W : ℕ+) (S : Type → Type) (F : Type) [Field F] [ProvableType S] where
-  /--
-    Add some witnessed variable to the context
-  -/
-  | Witness : CellOffset W S → (compute : Unit → F) → TableConstraintOperation W S F
+inductive Cell (W: ℕ+) (S : Type → Type) [ProvableType S] where
+  | input : CellOffset W S → Cell W S
+  | aux : ℕ → Cell W S
 
-  /--
-    Witness a fresh variable for each cell in the row at some offset `off` in the trace
-  -/
-  | GetRow : (off : Fin W) → TableConstraintOperation W S F
+instance : Repr (Cell W S) where
+  reprPrec cell _ := match cell with
+    | .input off => ".input " ++ reprStr off
+    | .aux i => ".aux " ++ reprStr i
 
-  /--
-    Allocate a subcircuit in the trace
-  -/
-  | Allocate: {n: ℕ} → SubCircuit F n → TableConstraintOperation W S F
+/--
+Mapping between cell offsets in the table and variable indices.
 
-  /--
-    Assign a variable to a cell in the trace
-  -/
-  | Assign : Variable F → CellOffset W S → TableConstraintOperation W S F
+The mapping must maintain the invariant that each variable is assigned to at most one cell.
+On the other hand, a cell can be assigned zero, one or more variables.
+-/
+structure CellAssignment (W: ℕ+) (S : Type → Type) [ProvableType S] where
+  offset : ℕ -- number of variables
+  aux_length : ℕ -- maximum number of auxiliary cells (i.e. those not part of the input/output layout)
+
+  /-- every variable is assigned to exactly one cell in the trace -/
+  vars : Vector (Cell W S) offset
```

Comment:
we redefine a `CellAssignment` to be a vector of variable assignments, rather than a function.

now, _all_ variables get assigned, and they can be either assigned to window cells or "auxiliary" cells, which hold any internal vars of our circuit. the idea is that some post processing could automatically add those aux vars at the end of the last row of each window